multi: match fee rates and smart fees

chappjc opened the pull request:
Resolving https://github.com/decred/dcrdex/issues/502

### Purpose

Smart fee rates from server provided to clients for their swap contract txns.

### Details

**Server asset fee estimation**
- The server's asset backends provide an optimal fee rate (e.g. estimatesmartfee with dcr/btc/etc.). `asset.Backend` interface method `FeeRate() (uint64, error)` is added.
- Fee rate is in atoms/byte, but we could consider atoms/KB for increased precision.

**Fee rate communication**
- Instead of the `"feeRate"` asset config parameter, it is now `"maxFeeRate"`, which indicates how much clients are required to reserve in their order-funding coins to cover the worst case fees at swap time.
- The server-originating `'match'` request payload (`msgjson.Match`) now includes `"feeratebase"` and `"feeratequote"`.  The client only cares about one presently, the one that applies to their swap contract txns, but the other could be used for both auditing the counterparty contract txn and possibly their own redeem txn.  The `msgjson.Match` serialization=>signature is updated to include these fields.  The `dex/order.Match` type now also includes `FeeRateBase` and `FeeRateQuote` fields, but these are _not_ included in the serialization=>matchID.

**Server-side (Swapper) contract fee rate handling**
- The Swapper computes required fee rates on start of negotiation (after market matching and before sending `'match'` requests). If fee estimation fails or returns a value higher than the configured maxFeeRate, it uses maxFeeRate.
- The Swapper checks each contract txn's fee rate, ensuring it is at least the required fee rate for the relevant asset as set in the `matchTracker.Match.FeeRate{Base,Quote}`.
- The redemption txns' fee rates are not checked as this is the user's problem.

**Client-side contract fee rate handling**
- Use the provided fee rate from the `'match'` message.
- Since swap transactions batch several contracts, the highest fee rate across all of the batched swaps is used.
- Client assets estimate their own fee in `Redeem`, `Refund`, `PayFee`, and `Withdraw` internally (with a `FallbackFeeRate`).  `Swap` uses the fee rate provided by the server.
- **TODO**: Setting for fallback fee rate: https://github.com/decred/dcrdex/pull/505/commits/a78171bf14785d6bc63be5c31aa02477a5f378b8

**Transaction sizing and order funding**
- Several transaction sizing fixes, and a new `InitTxSizeBase` const that does not include inputs.
- Order funding had incorrectly computed the required funds for a chain of N swaps while double counting the already-included input of the first swap in the chain. 
    - Fix `RequiredFunds` (now `RequiredOrderFunds`) to account for this when adding inputs for the first txn in the chain, and perform all computations in integer instead of float.
    - Client exchange wallets' `FundOrder` method uses `RequiredOrderFunds` with the `InitTxSizeBase`.
- Server assets provide an `InitTxSizeBase` method.
- `Market` is updated to check required funds on new orders using `RequiredOrderFunds` and `InitTxSizeBase`.

**Server DB updates for contract fee rates**
- Store the base and quote rates in the `matches` table.
- Finally use the placeholder `takerSell` column, using the value to load the correct rate for `UserMatch.FeeRateSwap`.

**Update Swapper startup for the recently added match time field**
- The recent addition of `matchTracker.MatchTime` in https://github.com/decred/dcrdex/pull/438 requires reloading this time from `matchTracker.Match.Epoch.End()`.

chappjc commented:
COMPLETELY untested (aside from unit tests and CI), but the objectives are reached so please feel free to have a look.

chappjc commented:
https://github.com/decred/dcrd/pull/2232 testnet issues.

But I'm afraid to say I've not taken this for a spin yet, so there's likely many things I've neglected.

JoeGruffins commented on client/asset/btc/btc.go:
```diff
@@ -401,9 +408,25 @@ func (btc *ExchangeWallet) Balance() (*asset.Balance, error) {
+		return 0, fmt.Errorf("no fee rate available")
+	}
+	satPerKB, err := btcutil.NewAmount(*feeResult.FeeRate)
+	return uint64(satPerKB) / 1000, err
```
In @davecgh`s  words:
>It's almost never good practice to return non-zero values in the case errors, unless the value is specifically known to be good in that case (e.g. a reader that returns the number of bytes read until the error happened, etc).

chappjc replied:
That's good practice, and we are actually doing that because we know `NewAmount` does it (and that `0 / s == 0`).

JoeGruffins commented on client/asset/dcr/dcr.go:
```diff
@@ -397,15 +411,30 @@ func (dcr *ExchangeWallet) Balance() (*asset.Balance, error) {
+	if err != nil {
+		return 0, err
+	}
+	atomsPerKB, err := dcrutil.NewAmount(dcrPerKB)
+	return uint64(atomsPerKB) / 1000, err // dcrPerKB * 1e8 / 1e3
```
here too

chappjc commented on dex/networks/dcr/script.go:
```diff
@@ -44,31 +41,39 @@ const (
+
+	// InitTxSizeBase is the size of a standard serialized atomic swap
+	// initialization transaction with one change output and no inputs. MsgTx
+	// overhead is 4 bytes version + 4 bytes locktime + 4 bytes expiry + 3 bytes
+	// of varints for the number of transaction inputs (x2 for witness and
+	// prefix) and outputs. There is one P2SH output with a 23 byte pkScript,
```
Good catch! This was BTC docs, which does use 4 byes for version, but also the formula in MsgTxOverhead is misleading despite giving the correct number.

You're probably referring to the following description, which I'll just drop here for reference:

https://github.com/decred/dcrd/blob/01fd2de07837ed44c8b24c77323e90ff0f45f6f1/wire/msgtx.go#L759-L770

dnldd commented on dex/msgjson/types.go:
```diff
@@ -367,63 +367,80 @@ func (msg *Message) String() string {
-	// Init serialization is orderid (32) + matchid (32) + txid (probably 32) +
-	// vout (4) + contract (97 ish). Sum = 197
+	// Init serialization is orderid (32) + matchid (32) + coinid (probably 36)
+	// + contract (97 ish). Sum = 197
```
`coinid (max 36) + contract (max 97)` would be more definitive assuming the sizes there are the maximum possible sizes.

chappjc replied:
Unfortunately coinid could be 24 bytes or 240 bytes, just depends on how the asset represents it's funds.

chappjc commented on client/asset/btc/livetest/livetest.go:
```diff
@@ -280,7 +280,7 @@ func Run(t testKiller, newWallet WalletConstructor, address string, dexAsset *de
 	}
 
-	_, _, err = rig.alpha().Redeem(redemptions, dexAsset)
+	_, _, err = rig.alpha().Redeem(redemptions, dexAsset.MaxFeeRate /* TODO */)
```
Ah, use a fee estimate or fallback fee.  Good catch.

chappjc commented on client/core/core.go:
```diff
@@ -897,7 +897,11 @@ func (c *Core) loadWallet(dbWallet *db.Wallet) (*xcWallet, error) {
 	walletCfg := &asset.WalletConfig{
-		Account:  dbWallet.Account,
+		Account: dbWallet.Account,
+		// TODO: FallbackFeeRate:
+		// - Update the DB structure.
+		// - Allow it to be set on new wallet construction.
+		// - Allow it to be reconfigured.
```
This TODO can't wait, or at least can't default to 0.  Thoughts welcome.

buck54321 commented on client/asset/btc/btc.go:
```diff
@@ -422,8 +445,20 @@ func (btc *ExchangeWallet) Fund(value uint64, nfo *dex.Asset) (asset.Coins, erro
+	// swaps, can we make sure that the change outputs that fund further swaps
+	// P2WPKH or is this up to the wallet software???
+
+	// TODO 2: We are making legacy/non-segwit swap contracts pkScripts. Can we
+	// safely make them all P2WSH?
```
Probably? Haven't actually tested it, but as long as the client and server agree, I think it should work.

buck54321 commented on client/asset/dcr/dcr.go:
```diff
@@ -1146,12 +1178,19 @@ func (dcr *ExchangeWallet) Lock() error {
+		dcr.log.Errorf("PayFee: Unable to get optimal fee rate, using fallback %d: %v",
+			dcr.fallbackFeeRate, err)
+	}
+	// TODO: why not SendToAddress like in btc ExchangeWallet?
```
If I'm remembering correctly, it's because dcrwallet randomizes the change output index, so creating the coin would require a bit of extra work.

buck54321 commented on client/core/trade.go:
```diff
@@ -1001,6 +1022,10 @@ func (t *trackedTrade) processAudit(msgID uint64, audit *msgjson.Audit) error {
 		return errs.add("swap contract value %d was lower than expected %d", auditInfo.Coin().Value(), auditQty)
 	}
 
+	// TODO: Consider having the server supply a fee rate: If this is taker,
+	// apply fee rate to their swap; if this is maker, apply fee rate to their
+	// redeem.
+
```
Is this comment outdated? The server is supplying both fee rates, I think.

chappjc replied:
Mostly outdated, but in `processAudit`this still refers to possible checks that client can do on the counter party's swap txn.  I've revised the comment in `processAudit` to read as follows:

```go
// TODO: Consider having the server supply the contract txn's fee rate to
// improve the taker's audit with a check of the maker's contract fee rate.
// The server should be checking the fee rate, but the client should not
// trust it. The maker could also check the taker's contract txn fee rate,
// but their contract is already broadcasted, so the check is of less value
// as they can only wait for it to be mined to redeem it, in which case the
// fee rate no longer matters, or wait for the lock time to expire to refund.
```

chappjc commented on dex/order/match.go:
```diff
@@ -189,7 +193,7 @@ func (match *Match) ID() MatchID {
 	b := make([]byte, 0, 2*OrderIDSize+8+8)
 	b = appendOrderID(b, match.Taker)
-	b = appendOrderID(b, match.Maker)
+	b = appendOrderID(b, match.Maker) // this maker and taker may only be matched once
```
The MatchID is unique with this order pair since you won't ever see this pair of orders matched again.

chappjc commented on client/core/trade.go:
```diff
@@ -794,7 +810,11 @@ func (t *trackedTrade) redeemMatches(matches []*matchTracker) error {
 	// Send the transaction.
 	redeemWallet, redeemAsset := t.wallets.toWallet, t.wallets.toAsset // this is our redeem
-	coinIDs, outCoin, err := redeemWallet.Redeem(redemptions, redeemAsset)
+	feeRate, err := redeemWallet.FeeRate()
+	if err != nil {
+		feeRate = redeemAsset.MaxFeeRate // or redeemWallet.Info().DefaultFeeRate or a value from a server message?
+	}
+	coinIDs, outCoin, err := redeemWallet.Redeem(redemptions, feeRate)
```
That's an option.  I think because I'm not yet certain Redeem won't want to use a server-based fee or what.  I'm OK with having this be done internally at this point though.

chappjc commented on client/asset/btc/btc.go:
```diff
@@ -996,6 +1031,14 @@ func (btc *ExchangeWallet) Refund(coinID, contract dex.Bytes, nfo *dex.Asset) (d
 		return nil, fmt.Errorf("error extracting swap addresses: %v", err)
 	}
+
+	feeRate, err := btc.FeeRate()
+	if err != nil {
+		feeRate = btc.fallbackFeeRate
+		btc.log.Errorf("Refund: Unable to get optimal fee rate, using fallback %d: %v",
+			btc.fallbackFeeRate, err)
+	}
```
Sure. `FeeRateWithFallback` or similar?  I do want to have `FeeRate` not silently use fallback.

chappjc commented on client/asset/btc/btc.go:
```diff
@@ -422,8 +445,20 @@ func (btc *ExchangeWallet) Fund(value uint64, nfo *dex.Asset) (asset.Coins, erro
 	var spents []*output
 	fundingCoins := make(map[string]*compositeUTXO)
 
+	// TODO: We are calculating required funds assuming segwit (discounted
+	// witness), but existing utxos may not be witness and thus we should be
+	// counting total input bytes rather than vbytes. Also, even for the chained
+	// swaps, can we make sure that the change outputs that fund further swaps
+	// P2WPKH or is this up to the wallet software???
```
Poorly phrased with "bytes rather than vbytes".  But good point that the `input.VBytes()` method handles this.

I'll revise this comment with a note about generating p2wpkh outputs.

chappjc commented on server/asset/btc/btc_test.go:
```diff
@@ -301,6 +301,17 @@ func txOutID(txHash *chainhash.Hash, index uint32) string {
+const optimalFeeRate uint64 = 24
+
+func (testNode) EstimateSmartFee(confTarget int64, mode *btcjson.EstimateSmartFeeMode) (*btcjson.EstimateSmartFeeResult, error) {
+	optimalRate := float64(optimalFeeRate) * 1e-5
+	// fmt.Println((float64(optimalFeeRate)*1e-5)-0.00024)
```
BTW, the kind of thing that makes be super uncomfortable with floating point math: https://play.golang.org/p/Dhmxmp5OosM

The error is less than 1e-8, so it's going to work out to the same atoms/byte, but it still sucks.

chappjc commented:
@buck54321 Addressed your comments in https://github.com/decred/dcrdex/pull/505/commits/0f80a22c97012a0f0f15c2066efef589b0af9e87, https://github.com/decred/dcrdex/pull/505/commits/73c00f77f83035dcd45e17c88846927a6bf64a58, and https://github.com/decred/dcrdex/pull/505/commits/00d632a26ac60d775b01a5b3b43d1c9d40dbffb9.

chappjc commented:
Tested now that it's rebased on master, and it worked as expected.  However, fee estimation seems to not work on BTC simnet/regnet, so the server's configured fallback fee is sent to the client for BTC.

**dcrdex**

The server got the optimal fee rate of 10 atoms/byte and the fallback of 100 sat/byte:

```
[INF] MKT: Matching complete for market dcr_btc epoch 265603419: 1 matches (0 partial fills), 1 completed OK (not booked), 0 booked, 1 unbooked, 0 failed
[DBG] MKT: Negotiating 1 matches for epoch 265603419:6000
[DBG] SWAP: Optimal fee rate for "dcr": 10
[WRN] SWAP: Unable to determining optimal fee rate for "btc", using fallback of 100. Err: Insufficient data or no feerate found
```

All negotiation steps including fee rate checks succeeded:

```
[DBG] AUTH: Recorded order 4056844955cce70784bf6c7591b8b2f47abaf47771500e13947ce5a5886614f9 that has finished processing: user=05b1c1137a602501bed7c9791f4b20d5240e47584dd24ce724cbc7659b18f697, time=1593620524791, target=<nil>
[DBG] SWAP: handleRedeem: 'redeem' received from 05b1c1137a602501bed7c9791f4b20d5240e47584dd24ce724cbc7659b18f697 for match adc293ba766ac589b939c217a135ce7027c34ab9e85057e07890b010a30c7a9e, order 4056844955cce70784bf6c7591b8b
[DBG] SWAP: step(user=05b1c1137a602501bed7c9791f4b20d5240e47584dd24ce724cbc7659b18f697, match=adc293ba766ac589b939c217a135ce7027c34ab9e85057e07890b010a30c7a9e): ready for user 05b1c1137a602501bed7c9791f4b20d5240e47584dd24ce72
[DBG] SWAP: processRedeem: valid redemption received at 2020-07-01 16:22:04.803 +0000 UTC from 05b1c1137a602501bed7c9791f4b20d5240e47584dd24ce724cbc7659b18f697 for match adc293ba766ac589b939c217a135ce7027c34ab9e85057e07890b01
[DBG] SWAP: processRedeem: sending 'redeem' ack request to counterparty 05b1c1137a602501bed7c9791f4b20d5240e47584dd24ce724cbc7659b18f697 for match adc293ba766ac589b939c217a135ce7027c34ab9e85057e07890b010a30c7a9e
[DBG] AUTH: Recorded order 28c9bd49fa4d6769a455b10b9c6fd75566d59cd2c647c34b932e9cfb57a0a3fc that has finished processing: user=05b1c1137a602501bed7c9791f4b20d5240e47584dd24ce724cbc7659b18f697, time=1593620524808, target=<nil>
```


**dexc**

The client used the fee rates provided in the `'match'` message:

```
[INF] CORE: Broadcasted transaction with 1 swap contracts for order 28c9bd49fa4d6769a455b10b9c6fd75566d59cd2c647c34b932e9cfb57a0a3fc. Fee rate = 100. Receipts: [0baac477af6eca6e105c6e158a3f4aa253821172cff3fa0225ae0e78c9b68347:0]
...
[INF] CORE: Broadcasted transaction with 1 swap contracts for order 4056844955cce70784bf6c7591b8b2f47abaf47771500e13947ce5a5886614f9. Fee rate = 10. Receipts: [0b52c01391b5b35d7b91202b8de713a516f9808083a8214f9be1edee101094ca:0]
```

Since `client/asset.Wallet.Redeem(contract)` uses fee estimation internally now, BTC used the fallback fee and DCR used a fee estimate:

```
[WRN] CORE[btc]: Unable to get optimal fee rate, using fallback of 80: Insufficient data or no feerate found
```

chappjc commented on dex/calc/fees.go:
```diff
@@ -1,14 +1,29 @@
+// size of a standard swap with no inputs. The chained swap txns will be the
+// standard size as they will spend a previous swap's change output.
+// TODO: change the inputs when/if we eliminiate dex.Asset.SwapSize.
+func RequiredOrderFunds(swapVal, inputsSize, baseSwapSize uint64, coin *dex.Asset) uint64 {
+	maxSwaps := swapVal / coin.LotSize
+	if maxSwaps == 0 {
+		return math.MaxUint64 // caller messed up
+	}
+
+	baseBytes := maxSwaps * coin.SwapSize
+	// SwapSize already includes one input, replace the size of the first swap
+	// in the chain with the given size of the actual inputs + baseSwapSize.
+	firstSwapSize := inputsSize + baseSwapSize
+	totalBytes := baseBytes + firstSwapSize - coin.SwapSize // in this order because we are using unsigned integers
```
I was hoping to chat about this.  I actually thinking `SwapSize` may not need to be a server-set field, and then we could have `RequiredOrderFunds` take both `baseSwapSize` and `swapSize`.  Although I understand there's no guarantee that the client and server share the same version of the `dex/networks` packages (or use it at all) where these values are defined.  Still, the consequences of not funding enough are just the inability to submit orders rather than failed swaps.  Hmm, if there's a way to avoid adding more data to the messages that would be better, but it looks like you're right.

chappjc commented on client/asset/interface.go:
```diff
@@ -42,6 +43,9 @@ type WalletConfig struct {
 	// Account is the name of the wallet account. The account and wallet must
 	// already exist.
 	Account string
+	// FallbackFeeRate is fee rate used for withdraws, refunds, and DEX fee
+	// payment transactions when optimal fee estimation is unavailable.
+	FallbackFeeRate uint64
```
Thought about `ConfigOpts`, but I got the impression that was for wallet software (dcrwallet, bitcoind, etc.) settings rather than settings for the Go wallet implementation.  Will revisit this.

chappjc replied:
Anyway, punt as per matrix chat.  @itswisdomagain Maybe this could be a follow up for you.  There's a sort of obsolete `WalletInfo.DefaultFeeRate` now and `WalletConfig.FallbackFeeRate`.  The best way to clean this up depends on when, how, and where we want the fallback fee rate to be set by the user.

itswisdomagain replied:
Right. Will think up and share some ideas and put up a PR once an issue for this is put up.

chappjc commented on dex/networks/btc/script.go:
```diff
@@ -516,15 +534,26 @@ func InputInfo(pkScript, redeemScript []byte, chainParams *chaincfg.Params) (*Sp
+		// If the redeem script is p2ms, there is a leading OP_0, but we do not
+		// know if it is actually multisig since M and N could both be 1. TODO:
+		// change this test to use the redeem script's type!
+		if scriptAddrs.NRequired > 1 /* not good enough, could be 1-of-1! */ {
+			sigScriptSize++ // P2SM's leading OP_0
+		}
```
Ah, yes `ParseScriptType` already checked the redeem script this way.  I was thinking it would require another parse of the redeem script, but it turns out it's already done.

chappjc commented:
Tested on testnet now, also with the `SwapSizeBase` in the `'config'` response.  Fee estimation worked on both DCR and BTC.

C0: https://tbtc.bitaps.com/b31b4f861b8e97728559e88ea66ad9c964c417546db5e5fcec1e47e72bcce1ec
R0: https://testnet.dcrdata.org/tx/9ec8062b3ea138beb1bac95765a4b9e303702ab4fff7e5252770b11ff941c911/in/0

C1: https://testnet.dcrdata.org/tx/2e6a113d513d0e99b1d6e2ce02e465f6c9841548a8954a1156e904acb113c9b6
R1: https://tbtc.bitaps.com/0e02ff305cea4d71ec5afbd67c1a1abee7e13358446407d489f94c4f05bd2b0d/input/0
